app/models/repository/git_lab.rb

```diff
@@ -0,0 +1,229 @@
+  end
+
+  def clone_path
+    self.url
+  end
+
+  def clone_host
+    p = parse(clone_url)
+    return p[:host]
+  end
+
+  def clone_protocol_http?
+    # Possible valid values (via http://git-scm.com/book/ch4-1.html):
+    #  ssh://user@server/project.git
+    #  user@server:project.git
+    #  server:project.git
+    # For simplicity we just assume if it's not HTTP(S), then it's SSH.
+    clone_url.match(/^http/)
+  end
+
+  # Hook into Repository.fetch_changesets to also run 'git fetch'.
+  def fetch_changesets
+    # ensure we don't fetch twice during the same request
+    return if @already_fetched
+    @already_fetched = true
+
+    puts "Calling fetch changesets on #{clone_path}"
+    # runs git fetch
+    self.fetch
+    super
+  end
+
+  # Override default_branch to fetch, otherwise caching problems in
+  # find_project_repository prevent Repository::Git#fetch_changesets from running.
+  #
+  # Ideally this would only be run for RepositoriesController#show.
+  def default_branch
+    if self.branches == [] && self.project.active? && Setting.autofetch_changesets?
+      # git_adapter#branches caches @branches incorrectly, reset it
+      scm.instance_variable_set :@branches, nil
+      # NB: fetch_changesets is idemptotent during a given request, so OK to call it 2x
+      self.fetch_changesets
+    end
+    super
+  end
+
+  # called in before_validate handler, sets form errors
+  def initialize_clone
+    # avoids crash in RepositoriesController#destroy
+    return unless attributes["extra_info"]["extra_clone_url"]
+
+    p = parse(attributes["extra_info"]["extra_clone_url"])
+    self.identifier = p[:identifier] if identifier.empty?
+
+    base_path = Setting.plugin_redmine_git_remote['git_remote_repo_clone_path']
+    base_path = base_path + "/" unless base_path.end_with?("/")
+
+    self.url = base_path + p[:path] if url.empty?
+
+    err = ensure_possibly_empty_clone_exists
+    errors.add :extra_clone_url, err if err
+  end
+
+  # equality check ignoring trailing whitespace and slashes
+  def two_remotes_equal(a, b)
+    a.chomp.gsub(/\/$/, '') == b.chomp.gsub(/\/$/, '')
+  end
+
+  def ensure_possibly_empty_clone_exists
+
+    if clone_protocol_http?
+      url = extra_clone_url.to_s.split('://')[1]
+      git_url = "https://#{clone_username}:#{clone_token}@#{url}"
+      cmd = "git ls-remote -h #{git_url}"
+      unless system(cmd)
+        return "#{clone_url} is not a valid remote."
+      end
+    else
+      unless system "git", "ls-remote", "-h", clone_url
+        return "#{clone_url} is not a valid remote."
+      end
+    end
+
+    puts "Clone path #{clone_path}"
+
+    if Dir.exists?(clone_path)
+      if clone_protocol_http?
+        existing_repo_remote, status = RedmineGitRemote::PoorMansCapture3::capture2(
+          "git", "--git-dir", clone_path, "config", "--get", "remote.origin.url")
+
+      else
+        cmd = "git --git-dir #{clone_path} config --get #{clone_username}:#{clone_token}@remote.origin.url"
```

Comment:
We do not want to put the credentials in the URL.
Also, for GitLab repositories we want to use an access token.
They way it is currently done in the GitRemote class.